// @indent/webhook/dist/signature/sign.test.js
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sign_1 = require("./sign");
var exampleRequest = {
    method: 'POST',
    host: 'example123.execute-api.us-west-2.amazonaws.com',
    path: '/webhook/dev',
    headers: {
        Host: 'example123.execute-api.us-west-2.amazonaws.com',
        'X-Indent-Timestamp': '2020-05-02T06:10:30Z'
    },
    body: { events: [{ hello: true }] }
};
describe('sign', function () {
    test('basic', function () {
        return expect(sign_1.sign('secret_123', {
            method: 'POST',
            host: 'example123.execute-api.us-west-2.amazonaws.com',
            path: '/webhook/dev',
            headers: {
                Host: 'example123.execute-api.us-west-2.amazonaws.com',
                'X-Indent-Timestamp': '2020-05-02T06:10:30Z'
            },
            body: { events: [{ hello: true }] }
        })).toEqual('5776de99671e05ca680fe2f33ac311c496969fd782d01c43793a21ac7280b1de');
    });
    test('different secrets should fail', function () {
        return expect(sign_1.sign('secret1', exampleRequest)).not.toEqual(sign_1.sign('secret2', exampleRequest));
    });
});
describe('time', function () {
    test('invalid timestamp', function () {
        return expect(sign_1.sign('secret1', exampleRequest)).toEqual('bf112e8bb28cbadfe4ec87f9f6e8430859a6c9d78b95e0dc9c0b3a32a82b173a');
    });
    test('different times should fail', function () {
        return expect(sign_1.sign('secret1', exampleRequest)).not.toEqual(sign_1.sign('secret1', tslib_1.__assign(tslib_1.__assign({}, exampleRequest), { timestamp: '2020-05-02T06:10:31Z' })));
    });
});
