// @indent/base-integration/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StatusCode = void 0;
var StatusCode;
(function (StatusCode) {
    StatusCode[StatusCode["OK"] = 0] = "OK";
    StatusCode[StatusCode["CANCELLED"] = 1] = "CANCELLED";
    StatusCode[StatusCode["UNKNOWN"] = 2] = "UNKNOWN";
    StatusCode[StatusCode["INVALID_ARGUMENT"] = 3] = "INVALID_ARGUMENT";
    StatusCode[StatusCode["DEADLINE_EXCEEDED"] = 4] = "DEADLINE_EXCEEDED";
    StatusCode[StatusCode["NOT_FOUND"] = 5] = "NOT_FOUND";
    StatusCode[StatusCode["ALREADY_EXISTS"] = 6] = "ALREADY_EXISTS";
    StatusCode[StatusCode["PERMISSION_DENIED"] = 7] = "PERMISSION_DENIED";
    StatusCode[StatusCode["RESOURCE_EXHAUSTED"] = 8] = "RESOURCE_EXHAUSTED";
    StatusCode[StatusCode["FAILED_PRECONDITION"] = 9] = "FAILED_PRECONDITION";
    StatusCode[StatusCode["ABORTED"] = 10] = "ABORTED";
    StatusCode[StatusCode["OUT_OF_RANGE"] = 11] = "OUT_OF_RANGE";
    StatusCode[StatusCode["UNIMPLEMENTED"] = 12] = "UNIMPLEMENTED";
    StatusCode[StatusCode["INTERNAL"] = 13] = "INTERNAL";
    StatusCode[StatusCode["UNAVAILABLE"] = 14] = "UNAVAILABLE";
    StatusCode[StatusCode["DATA_LOSS"] = 15] = "DATA_LOSS";
    StatusCode[StatusCode["UNAUTHENTICATED"] = 16] = "UNAUTHENTICATED";
})(StatusCode = exports.StatusCode || (exports.StatusCode = {}));
